File: graph/course_schedule2.py
from typing import List
from collections import deque

class Solution:
    # dfs
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        prereq = {c:[] for c in range(numCourses)}
        for crs, pre in prerequisites:
            prereq[crs].append(pre)
            
        output = []
        visit, cycle = set(), set()
        
        def dfs(crs):
            if crs in cycle:
                return False
            if crs in visit:
                return True
            cycle.add(crs)
            for pre in prereq[crs]:
                if dfs(pre) == False:
                    return False
            cycle.remove(crs)
            visit.add(crs)
            output.append(crs)
            return True
        
        for c in range(numCourses):
            if dfs(c) == False:
                return []
            
        return output
    
    # bfs
    def findOrder2(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        edges = [[] for i in range(numCourses)]
        degrees = [0]*numCourses
        output = []
        q = deque()

        for cur, preCur in prerequisites:
            edges[preCur].append(cur)
            degrees[cur] += 1

        for cur, degree in enumerate(degrees):
            if degree == 0:
                q.append(cur)

        while q:
            cur = q.popleft()
            output.append(cur)
            for preCur in edges[cur]:
                degrees[preCur] -= 1
                if degrees[preCur] == 0:
                    q.append(preCur)
        
        return output if len(output) == numCourses else []

File: graph/course_schedule2.java
public class Solution {
    public int[] findOrder1(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> prereq = new HashMap<>();
        for(int i = 0; i < numCourses; i++){
            prereq.put(i, new ArrayList<>());
        }
        for (int[] pair : prerequisites) {
            prereq.get(pair[0]).add(pair[1]);
        }

        List<Integer> output = new ArrayList();
        Set<Integer> cycle = new HashSet<>();
        Set<Integer> visit = new HashSet<>();

        for(int c = 0; c < numCourses; c++){
            if(!dfs(c, cycle, visit, prereq, output)){
                return new int[]{};
            }
        }

        int[] outputArr = new int[output.size()];
        for(int i = 0; i < output.size(); i++){
            outputArr[i] = output.get(i);
        }

        return outputArr;

    }

    public boolean dfs(int crs, Set<Integer> cycle, Set<Integer> visit, Map<Integer, List<Integer>> prereq, List output){
        if(cycle.contains(crs)){
            return false;
        }
        if(visit.contains(crs)){
            return true;
        }
        cycle.add(crs);
        for(int pre : prereq.get(crs)){
            if(!dfs(pre, cycle, visit, prereq, output)){
                return false;
            }
        }
        cycle.remove(crs);
        visit.add(crs);
        output.add(crs);
        return true;
    }

    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> edges = new ArrayList<>();
        for(int i = 0; i < numCourses; i++){
            edges.add(new ArrayList<>());
        }
        int[] degrees = new int[numCourses];
        List<Integer> output = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            edges.add(new ArrayList<>());
        }

        for (int[] prerequisite : prerequisites) {
            int cur = prerequisite[0];
            int preCur = prerequisite[1];
            edges.get(preCur).add(cur);
            degrees[cur]++;
        }

        for (int i = 0; i < numCourses; i++) {
            if (degrees[i] == 0) {
                q.offer(i);
            }
        }

        while (!q.isEmpty()) {
            int cur = q.poll();
            output.add(cur);
            for (int preCur : edges.get(cur)) {
                degrees[preCur]--;
                if (degrees[preCur] == 0) {
                    q.offer(preCur);
                }
            }
        }

        if (output.size() == numCourses) {
            int[] result = new int[numCourses];
            for (int i = 0; i < numCourses; i++) {
                result[i] = output.get(i);
            }
            return result;
        } else {
            return new int[]{};
        }
    }
}
